### DISCUSSION TITLE
Theme Settings

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#2874] [General]
I just noticed a behavioral "breaking" change in the 4.0.0 release which I would like to explain in advance so that it is not a surprise...

Since version 2.x, Oqtane has supported the concept of Theme Settings. Essentially this feature allows you to provide an Admin UI with your theme so that an end-user can customize the theme to meet their needs. The way to access Theme Settings was via the Control Panel and Page Edit button. This would display the standard Edit Page UI modal and an additional tab would be displayed if the page was using a Theme which supported Theme Settings. 

The challenge is that the Edit Page UI is a shared component - meaning it can be accessed in multiple ways. One way is via the Control Panel as I described above, and the other way is via the Page Management UI in the Admin Panel where an Administrator is able to view and manage a list of all pages in the site. In the past, the Edit Page UI contained some non-intuitive logic to deal with the different ways it could be accessed. When called from the Control Panel it expected a "cp=" querystring parameter to be passed, and when called from Page Management it expected an "id=" querystring parameter to be passed. In version 4.0.0 this was consolidated so that all access paths rely on the standard "id=" querystring parameter... and it uses the newer ReturnUrl concept to retain state so that it can redirect to the appropriate location on Save/Cancel. 

The problem is that legacy Theme Settings components which were included with Themes assumed that they could utilize the PageState.Page.Settings cascading parameter to access the page settings. However it is important to understand the PageState.Page.Settings collection contains items for the current page - not the page related to the "id=" querystring parameter. In order to workaround this issue in the past, the Theme Settings were not displayed in the Edit Page UI in the scenario where the user was managing pages via Page Management. However this was an artificial limitation and was only required because Theme Settings components were making invalid assumptions in their implementation.

In version 4.0.0, Theme Settings will be manageable from both the Control Panel and Page Management. But this means that Theme Settings components included in custom Themes will need to be aware that they need to explicitly load the Page Settings related to the "id=" querystring and NOT rely on the PageState.Page.Settings cascading parameter.

the logic for loading should be:
```
   private int _pageId = -1;
   private string _mySetting = "";

    protected override async Task OnInitializedAsync()
    {
       if (PageState.QueryString.ContainsKey("id"))
        {
            _pageId = int.Parse(PageState.QueryString["id"]);
            var settings = await SettingService.GetPageSettingsAsync(_pageId);
            _mySetting = SettingService.GetSetting(settings, GetType().Namespace + ":MySetting", "");
        }
    }
```

and saving should be:

```
       if (_pageId != -1)
      {
           var settings = await SettingService.GetPageSettingsAsync(_pageId);
           settings = SettingService.SetSetting(settings, GetType().Namespace + ":MySetting", _mySetting);
            await SettingService.UpdatePageSettingsAsync(settings, _pageId);
      }
```

@leigh-pointer I am guessing this might affect your Rocket Fuel theme... so FYI.

**Comment by leigh-pointer** [comment]
@sbwalker thanks for the info.  I will need to update the theme, and Templates

**Comment by sbwalker** [comment]
@leigh-pointer one other thing to keep in mind is that if you want your theme to support Oqtane 3 & 4 then you will need to handle both the "cp=" (legacy from control panel) and "id=" (new) querystring parameters
